File: intern/41.first-missing-positive.js
/*
 * @lc app=leetcode id=41 lang=javascript
 *
 * [41] First Missing Positive
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number}
 */
var firstMissingPositive = function (nums) {
  for (let i = 0; i < nums.length; i++) {
    while (nums[i] !== i + 1 && nums[i] <= nums.length && nums[i] > 0) {
      const anotherIndex = nums[i] - 1;
      if (nums[anotherIndex] === nums[i]) {
        nums[i] = undefined;
        break;
      } else {
        [nums[i], nums[anotherIndex]] = [nums[anotherIndex], nums[i]];
      }
    }
  }

  for (let i = 0; i < nums.length; i++) {
    if (nums[i] !== i + 1) {
      return i + 1;
    }
  }

  return nums.length + 1;
};
// @lc code=end

File: prev-2024/41.first-missing-positive.cpp
/*
 * @lc app=leetcode id=41 lang=cpp
 *
 * [41] First Missing Positive
 */
#include <cmath>
#include <vector>

using std::vector;

// @lc code=start
class Solution {
 public:
  int firstMissingPositive(vector<int>& nums) {
    int nextPositionIndex = 0;

    for (int i = 0; i < nums.size(); i++) {
      if (nums[i] > 0) {
        nums[nextPositionIndex] = nums[i];
        nextPositionIndex++;
      }
    }

    const int validLength = nextPositionIndex;
    for (int i = 0; i < validLength; i++) {
      int index = std::abs(nums[i]) - 1;
      if (index < validLength && nums[index] > 0) {
        nums[index] = -nums[index];
      }
    }

    for (int i = 0; i < validLength; i++) {
      if (nums[i] > 0) {
        return i + 1;
      }
    }

    return validLength + 1;
  }
};
// @lc code=end

int main() {
  Solution sol;
  vector<int> vec({7, 8, 9, 11, 12});
  sol.firstMissingPositive(vec);
}